gk/fib.c

```diff
@@ -0,0 +1,1326 @@
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a gateway fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
+			}
+
+			if (ip_prefix_fib->action == GK_FWD_GATEWAY_FRONT_NET)
+				neigh_fib = gk_conf->neigh_fib_front;
+			else
+				neigh_fib = gk_conf->neigh_fib_back;
+
+			neigh_ht = &neigh_fib->u.neigh;
+			ret = rte_hash_del_key(neigh_ht->hash_table,
+				&gw_addr->ip.v4.s_addr);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to delete the Ethernet cached header for the IP prefix %s\n",
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * TODO Add concurrency control to
+			 * reset the Ethernet cache entry.
+			 */
+			eth_cache->stale = false;
+			memset(&eth_cache->eth_hdr, 0,
+				sizeof(eth_cache->eth_hdr));
+			eth_cache->ref_cnt = 0;
+		} else if (gw_addr->proto == ETHER_TYPE_IPv6 &&
+				memcmp(gw_addr->ip.v6.s6_addr,
+				ip_prefix_addr.ip.v6.s6_addr,
+				sizeof(gw_addr->ip.v6) == 0)) {
+			/*
+			 * The Ethernet header cache entry is still
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a gateway fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
+			}
+
+			if (ip_prefix_fib->action == GK_FWD_GATEWAY_FRONT_NET)
+				neigh_fib = gk_conf->neigh6_fib_front;
+			else
+				neigh_fib = gk_conf->neigh6_fib_back;
+
+			neigh_ht = &neigh_fib->u.neigh6;
+			ret = rte_hash_del_key(neigh_ht->hash_table,
+				gw_addr->ip.v6.s6_addr);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to delete the Ethernet cached header for the IP prefix %s\n",
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * TODO Add concurrency control to
+			 * reset the Ethernet cache entry.
+			 */
+			eth_cache->stale = false;
+			memset(&eth_cache->eth_hdr, 0,
+				sizeof(eth_cache->eth_hdr));
+			eth_cache->ref_cnt = 0;
+		}
+
+		break;
+	}
+
+	case GK_DROP:
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		/* FALLTHROUGH */
+	default:
+		return -1;
+		break;
+	}
+
+	/*
+	 * If the updated FIB entry doesn't have an action GK_FWD_GRANTOR,
+	 * we don't need to flush the flow tables at all.
+	 */
+	if (ip_prefix_fib->action == GK_FWD_GRANTOR) {
+
+		rte_atomic16_init(&ip_prefix_fib->num_updated_instances);
+
+		/* Send the FIB entry to the GK mailboxes. */
+		for (i = 0; i < gk_conf->num_lcores; i++) {
+			notify_flow_flush(ip_prefix_fib,
+				&gk_conf->instances[i]);
+		}
+
+		/*
+		 * Wait until all the GK instances
+		 * finish to flush their flow tables.
+		 */
+		while (rte_atomic16_read(&ip_prefix_fib->num_updated_instances)
+			!= gk_conf->num_lcores);
```

Comment:
Hi @cjdoucette,

According to your argument, If there's a bug here, there's a bug in DPDK since it's how DPDK reads atomic values.

I don't really know the reason why it should work the way it is, but my best _guess_ is that the fact that atomic integers are structs, the compilers knows that it must dereference it and not cache the value in a register.